x/interchainstaking/keeper/msg_server.go

```diff
@@ -28,6 +35,313 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	if k.Keeper.GetGovAuthority() != msg.Authority {
+		return &types.MsgRegisterZoneResponse{},
+			govtypes.ErrInvalidSigner.Wrapf(
+				"invalid authority: expected %s, got %s",
+				k.Keeper.GetGovAuthority(), msg.Authority,
+			)
+	}
+
+	// get chain id from connection
+	chainID, err := k.GetChainID(ctx, msg.ConnectionID)
+	if err != nil {
+		return &types.MsgRegisterZoneResponse{}, fmt.Errorf("unable to obtain chain id: %w", err)
+	}
+
+	// if subzone
+	if msg.SubzoneInfo != nil {
+		if chainID != msg.SubzoneInfo.BaseChainID {
+			return &types.MsgRegisterZoneResponse{}, fmt.Errorf("incorrect ID \"%s\" for subzone \"%s\"", chainID, msg.SubzoneInfo.BaseChainID)
+		}
+
+		// get zone
+		baseZone, found = k.GetZone(ctx, msg.SubzoneInfo.BaseChainID)
+		if !found {
+			return &types.MsgRegisterZoneResponse{}, fmt.Errorf("unable to find base chain \"%s\" for subzone \"%s\"", chainID, msg.SubzoneInfo.BaseChainID)
+		}
+
+		// set chainID to be specified unique ID
+		chainID = msg.SubzoneInfo.ChainID
+	}
+
+	// get zone
+	_, found = k.GetZone(ctx, chainID)
+	if found {
+		return &types.MsgRegisterZoneResponse{}, fmt.Errorf("invalid chain id, zone for \"%s\" already registered", chainID)
+	}
+
+	connection, found := k.IBCKeeper.ConnectionKeeper.GetConnection(ctx, msg.ConnectionID)
+	if !found {
+		return &types.MsgRegisterZoneResponse{}, errors.New("unable to fetch connection")
+	}
+
+	clientState, found := k.IBCKeeper.ClientKeeper.GetClientState(ctx, connection.ClientId)
+	if !found {
+		return &types.MsgRegisterZoneResponse{}, errors.New("unable to fetch client state")
+	}
+
+	tmClientState, ok := clientState.(*tmclienttypes.ClientState)
+	if !ok {
+		return &types.MsgRegisterZoneResponse{}, errors.New("error unmarshaling client state")
+	}
+
+	if tmClientState.Status(ctx, k.IBCKeeper.ClientKeeper.ClientStore(ctx, connection.ClientId), k.IBCKeeper.Codec()) != ibcexported.Active {
+		return &types.MsgRegisterZoneResponse{}, errors.New("client state is not active")
+	}
+
+	zone := &types.Zone{
+		ChainId:            chainID,
+		ConnectionId:       msg.ConnectionID,
+		LocalDenom:         msg.LocalDenom,
+		BaseDenom:          msg.BaseDenom,
+		AccountPrefix:      msg.AccountPrefix,
+		RedemptionRate:     sdk.NewDec(1),
+		LastRedemptionRate: sdk.NewDec(1),
+		UnbondingEnabled:   msg.UnbondingEnabled,
+		ReturnToSender:     msg.ReturnToSender,
+		LiquidityModule:    msg.LiquidityModule,
+		DepositsEnabled:    msg.DepositsEnabled,
+		Decimals:           msg.Decimals,
+		UnbondingPeriod:    int64(tmClientState.UnbondingPeriod),
+		MessagesPerTx:      msg.MessagesPerTx,
+		Is_118:             msg.Is_118,
+		SubzoneInfo:        msg.SubzoneInfo,
+	}
+
+	// verify subzone if setting
+	if zone.IsSubzone() {
+		if err := types.ValidateSubzoneForBasezone(*zone, baseZone); err != nil {
+			return &types.MsgRegisterZoneResponse{}, err
+		}
+	}
+
+	k.SetZone(ctx, zone)
+
+	// generate deposit account
+	portOwner := chainID + ".deposit"
```

Comment:
Can we use the constants here please :)

https://github.com/ingenuity-build/quicksilver/blob/388cc8a5ea65a685d905c5790c5cd7f35303da69/x/interchainstaking/types/keys.go#L38